### PR TITLE
internal/server/singleprocess/state: Add ability to prune jobs and deployments

### PR DESCRIPTION
**Pull request by evanphx**
This limits deployments and jobs to 10,000 records each. The code only prunes the data from the memory indexes, it leaves in the disk.

The pruning of data is done on a regular interval rather than inline during data inserts (ie, it's like a typical garbage collection process). The pruning logic is shared between app records and jobs as well.

Pruning the data from disk would come later.

### CODE REVIEW

**Review by mitchellh — COMMENTED** [review]
* I like only pruning the indexes. That way if we mess anything up we aren't deleting _real data_. 
* I'd introduce total record counts as package vars to avoid the iteration in jobs since that seems pretty easy. 
* For the app ops, I feel like we should just use the cursor at all times rather than `ForEach`. The existence of the secondary unmarshal, index, seq number tracking, etc. logic makes me nervous. Seems easy to introduce a bug that only exists in one or the other "mode". 
* Similar to above with jobs, I'd add functions to `*Job` that is like `IsComplete() bool` rather than reproducing that state logic over and over where its easy to introduce a bug.

**Comment by evanphx** [comment]
1. No prob adding total record counts 👍 
2. I'll dry up that iteration code 👍 
3. Totes, another good drying 👍

File: internal/server/singleprocess/state/app_operation.go [hunk 1]
```diff
@@ -50,6 +61,9 @@ type appOperation struct {
 	seq map[string]map[string]*uint64
 }
 
+// By default, we allow 10000 records
+const DefaultMaximumIndexedRecords = 10000
```
Comment (mitchellh) [inline internal/server/singleprocess/state/app_operation.go:65]: I don't think we use this anywhere.

If we do, I don't mean to get too wordy, but I think we should try to do like `DefaultAppOperationMaxIndexedRecords` (ugh such Java), but just to be clear that this is for app ops only?

If we intend this to be for _anything_ this is fine but it should probably go in a diff file.

File: internal/server/singleprocess/state/prune.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// pruneOld uses the types in op to scan the table indicated and prune old records.
+// The op's check function can allow the process to skip records that shouldn't be
+// pruned regardless of their age.
+func pruneOld(memTxn *memdb.Txn, op pruneOp) (int, error) {
+	op.lock.Lock()
+
+	// Easy enough, just exit if we haven't hit the maximum
+	if *op.cur < op.max {
```
Comment (mitchellh) [inline internal/server/singleprocess/state/prune.go:25]: We should protect against `op.cur == nil`. One suggestion is maybe just add a `Validate` method to `pruneOp` and use ozzo-validate or something to just do some really basic checking across the board.

File: internal/server/singleprocess/state/state.go [hunk 3]
```diff
@@ -121,6 +132,39 @@ func (s *State) Close() error {
...
+		}
+
+		records += cnt
+	}
+
+	s.log.Debug("Finished pruning data",
+		"removed-jobs", jobs,
+		"removed-records", records,
+		"op-tables", len(pruneFns),
+	)
```
Comment (mitchellh) [inline internal/server/singleprocess/state/state.go:161]: Might be useful to dump this per-function maybe as a trace so we can see prunes by table.

File: internal/server/singleprocess/state/prune.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	indexArgs    []interface{}
+	max          int
+	cur          *int
+	check        func(val interface{}) bool
+}
+
+// pruneOld uses the types in op to scan the table indicated and prune old records.
+// The op's check function can allow the process to skip records that shouldn't be
+// pruned regardless of their age.
+func pruneOld(memTxn *memdb.Txn, op pruneOp) (int, error) {
```
Comment (mitchellh) [inline internal/server/singleprocess/state/prune.go:21]: I know you test the callers of this, but since this is such an important func and a really genuinely nice abstraction, I'd feel better and recommend unit testing this specific func too.

**Review by mitchellh — COMMENTED** [review]
This did end up really nice. Great job. A few non-trivial comments but this is great stuff.

File: internal/server/singleprocess/state/app_operation.go (hunk 1 above)
Comment (evanphx) [reply internal/server/singleprocess/state/app_operation.go]: I just removed it for now since we're not using it.

File: internal/server/singleprocess/state/prune.go (hunk 2 above)
Comment (evanphx) [reply internal/server/singleprocess/state/prune.go]: Added Validate(), good call.

File: internal/server/singleprocess/state/state.go (hunk 3 above)
Comment (evanphx) [reply internal/server/singleprocess/state/state.go:162]: That's also reported now, per table.

File: internal/server/singleprocess/state/prune.go (hunk 4 above)
Comment (evanphx) [reply internal/server/singleprocess/state/prune.go:31]: Great call, found a bug while testing `pruneOld` in isolation.

**Comment by mitchellh** [comment]
If you rebase this on main the tests should be fixed 👍

**Review by mitchellh — APPROVED** [review]
